R/google_cloud_storage.R

```diff
@@ -0,0 +1,283 @@
+  id_col <- avoid_conflict(colnames(df), "id")
+  # copy internal exp.file.id to the id column.
+  df[[id_col]] <- df[["exp.file.id"]]
+  # drop internal column and move the id column to the very beginning.
+  df %>% dplyr::select(!!rlang::sym(id_col), dplyr::everything(), -exp.file.id)
+}
+
+#'API that search files by search keyword then imports multiple same structure CSV files and merge it to a single data frame
+#'
+#'For col_types parameter, by default it forces character to make sure that merging the CSV based data frames doesn't error out due to column data types mismatch.
+# Once the data frames merging is done, readr::type_convert is called from Exploratory Desktop to restore the column data types.
+
+#'@export
+searchAndGetCSVFilesFromGoogleCloudStorage <- function(bucket = "", folder = "", search_keyword = "", for_preview = FALSE, delim, quote = '"',
+                                       escape_backslash = FALSE, escape_double = TRUE,
+                                       col_names = TRUE, col_types = readr::cols(.default = readr::col_character()),
+                                       locale = readr::default_locale(),
+                                       na = c("", "NA"), quoted_na = TRUE,
+                                       comment = "", trim_ws = FALSE,
+                                       skip = 0, n_max = Inf, guess_max = min(1000, n_max),
+                                       progress = interactive()) {
+
+  # search condition is case insensitive. (ref: https://www.regular-expressions.info/modifiers.html, https://stackoverflow.com/questions/5671719/case-insensitive-search-of-a-list-in-r)
+  tryCatch({
+    files <- googleCloudStorageR::gcs_list_objects(bucket = bucket, detail= "more", prefix = folder, delimiter = "/") %>%
+      filter(str_detect(name, stringr::str_c("(?i)", search_keyword)))
+  }, error = function(e) {
+    if (stringr::str_detect(e$message, "http_404 The specified bucket does not exist")) {
+      # Looking for error that looks like "http_404 The specified bucket does not exist.".
+      # This seems to be returned when the bucket itself does not exist.
+      stop(paste0('EXP-DATASRC-18 :: ', jsonlite::toJSON(c(bucket)), ' :: The Google Cloud Storage bucket does not exist.'))
+    } else if (stringr::str_detect(e$message, "http_404 Unspecified error")) {
+      # Looking for error that looks like "http_404 Unspecified error".
+      # This seems to be returned when the file does not exist.
+      stop(paste0('EXP-DATASRC-19 :: ', jsonlite::toJSON(c(bucket)), ' :: There is no file in the Google Cloud Storage bucket that matches with the file name.'))
+    }
+    else {
+      stop(e)
+    }
+  })
+  if (nrow(files) == 0) {
+    stop(paste0('EXP-DATASRC-19 :: ', jsonlite::toJSON(bucket), ' :: There is no file in the Google Cloud Storage bucket that matches with the file name.')) # TODO: escape bucket name.
```

Comment:
Check if the result of toJSON is an array expression as expected.

Reply:
Yes, it's an array expression.

```r
> jsonlite::toJSON("bucket1")
["bucket1"] 
```